dask_expr/shuffle.py

```diff
@@ -462,6 +461,55 @@ def _layer(self):
         return toolz.merge(dsk1, dsk2, dsk3, dsk4)
 
 
+class P2PShuffle(SimpleShuffle):
+    """P2P worker-based shuffle implementation"""
+
+    lazy_hash_support = False
+
+    def _layer(self):
+        from distributed.shuffle._shuffle import (
+            ShuffleId,
+            barrier_key,
+            shuffle_barrier,
+            shuffle_transfer,
+            shuffle_unpack,
+        )
+
+        dsk = {}
+        token = self._name.split("-")[-1]
+        _barrier_key = barrier_key(ShuffleId(token))
+        name = "shuffle-transfer-" + token
+        transfer_keys = list()
+        parts_out = (
+            self._partitions if self._filtered else list(range(self.npartitions_out))
+        )
+        for i in range(self.frame.npartitions):
+            transfer_keys.append((name, i))
+            dsk[(name, i)] = (
+                shuffle_transfer,
+                (self.frame._name, i),
+                token,
+                i,
+                self.npartitions_out,
+                self.partitioning_index,
+                parts_out,
```

Comment:
`shuffle_transfer` currently expects `parts_out` to be a `set`. If that matters, let's convert to `set` here. If it doesn't matter, let's relax the typing in `dask/distributed`.

Reply:
Thanks for pointing this out. We can probably use a `set` if that's what `shuffle_transfer` expects.